### ISSUE TITLE
No Transparency when Menu Shader Pipeline is OFF and Menu Color Theme is not PLAIN

### ISSUE THREAD

**Issue by retro-wertz** [libretro/RetroArch#5001]
# First and foremost consider this:
- Only RetroArch bugs should be filed here. Not core bugs or game bugs
- This is not a forum or a help section, this is strictly developer oriented

## Description

while core or game is running, when "Menu Shader Pipeline is OFF" and using Menu Color Themes, going to Menu does not make it transparent anymore. 
Menu Transparency with color theme only happens if selecting a shader pipeline, OR when pipeline is OFF, menu color should be PLAIN which without color will blend colors from background and Menu Labels making it hard to read.

### Expected behavior

Im one of those that does not use menu shader pipelines. this is often OFF as i wanted background to be semi-transparent and with color so that changing  aspect ratios, game palette, etc effects can be seen immediately without the added stress from menu pipeline covering the image and/or labels blending into background.

Previous behavior only sets transparency when pipeline is OFF, changing transparency color also was doable without using pipeline.

turning off menu shader pipeline also allows showing custom background images.

OTHERS:
when Menu Color Theme is set to PLAIN,  selecting a Menu Shader Pipeline does not show.

Suggestion is that transparency should also work if menu shader pipeline is OFF(less stress to builtin or slow gpu's).

### Actual behavior
-This shows Menu Shader Pipeline at Ribbon, and Menu Color Theme set to PLAIN and pipeline is not showing. also labels are hard to read with this kind of transparency
![screenshot_2017-05-31_17-22-38](https://cloud.githubusercontent.com/assets/23315382/26625151/cf003ca4-4625-11e7-8b7d-2fb0efb9932d.png)

### Steps to reproduce the bug

1. [First step]
2. [Second step]
3. [and so on...]

### Bisect Results

first affected commit https://github.com/libretro/RetroArch/commit/57503051ea33c619a92d9f795708e710897a338b

### Version/Commit
You can find this information under Information/System Information

- RetroArch: [version/commit]

### Environment information

- OS:  Ubuntu 16.04
- Compiler: Ubuntu 5.4.0-6ubuntu1~16.04.4

**Comment by Kivutar** [comment]
We could add a plain pipeline that draw only the gradient.

**Comment by dankcushions** [comment]
i agree with @retro-wertz . in fact, IMO it should be the default, and the shader should only be displayed when no content is loaded.

when I used the XMB i managed to achieve this with configuration - I used a translucent .PNG as my wallpaper. it's important it is slightly darkened (ie, a black color with 50% alpha or something like that) so that the menu options and icons are clearly visible. mine was actually a slight gradient with the top and bottom the darkest. but yes, I like the suggestion as a start.

**Comment by Kivutar** [comment]
We also need to keep the code simple. For now, there is a lot of nested conditionals to know if we're going to display:

 * a gradient
 * an image
 * a dark mask for xmb
 * a white mask for materialui
 * an animation on top of a gradient

In addition to this, there are:

 - the opacity setting
 - the dynamic backgrounds feature
 - the pipeline that can be set to OFF
 - the gradient that can be set to plain

There are just too much options, and this is why this background code is a unintuitive mess.

**Comment by psyke83** [comment]
@Kivutar,

Isn't the problem here? https://github.com/libretro/RetroArch/blob/master/menu/drivers/xmb.c#L2580

That line seems redundant (and overrides the same setting a few lines above). I tried a quick compile with that line removed, and the menu becomes transparent when the shader pipeline is off and any colour scheme set (not just "plain").

**Comment by Kivutar** [comment]
If you comment this line, does it affect the fading animation when triggering the menu from the game?

**Comment by psyke83** [comment]
To be honest I'm not sure re: fading animation (I can't see any difference), but I do know there's a bad interaction with https://github.com/libretro/RetroArch/pull/5002 (if merged)

If #5002 is merged, this line will erroneously force the menu to stay transparent at all times when the shader pipeline is off: https://github.com/libretro/RetroArch/blob/master/menu/menu_driver.c#L546.

**Comment by psyke83** [comment]
OK, I've uploaded https://github.com/libretro/RetroArch/pull/5003/commits/65ab34702802dcc2938618f524a2ef78273ee7e9

That should complement the other pull: https://github.com/libretro/RetroArch/pull/5002

5002 improves transparency behaviour, ensuring that the menu stays opaque when there's no game loaded (as opposed to merely when there's no dummy core loaded).
5003 allows transparency when the shader pipeline off, and also fixes an issue that was causing background wallpaper rendering to make the menu transparent at the wrong time (similar to the issue that 5002 fixed for the shader pipelines).

This may affect the visual appearance of the background wallpaper (more opaque looking), however, so maybe there's a better way to fix that?

**Comment by Kivutar** [comment]
I tried your two patches, 5002 and 5003 together. The result is nice, but there are some issues when I toggle fullscreen.

**Comment by psyke83** [comment]
What issues in particular? I see that when you toggle fullscreen while the menu is active and game loaded, the menu becomes opaque until you toggle the menu. But... that happens without 5002 and 5003.

**Comment by Kivutar** [comment]
Yes it also happens without 5002 and 5003, but I don't think it was happenning before https://github.com/libretro/RetroArch/pull/4985

**Comment by psyke83** [comment]
I don't think that #4985 introduced a bug, but perhaps it did expose one. I'll see if I can hunt down the issue.

The fullscreen issue doesn't happen with or without all my patches when "Pause when menu activated" is off, so it's related to the pausing of the emulator.

**Comment by psyke83** [comment]
I just tested build 1.5.0 with all settings default except for Menu Color Theme set to Plain - the same fullscreen bug occurs. 

Removing this line fixes the issue: https://github.com/libretro/RetroArch/blob/master/gfx/video_driver.c#L1066

Will I submit another patch or merge it into an existing one?

**Comment by Kivutar** [comment]
Oh, thanks, you can merge it in an existing one. I will merge your patches when you tell me it's ready.

**Comment by psyke83** [comment]
I updated 5003 with the fullscreen fix. Both pull requests should be ready to merge as long as you're satisfied.

**Comment by Kivutar** [comment]
Ok thanks for your patches. We postpone the merging of these PR because we release 1.6.0 now. I will merge them just after the release.

**Comment by twinaphex** [comment]
Hi there, if you have patches which are ready right now, then go ahead and we can merge them in time for 1.6.0. I hope though that most of the bugs will be gone.

**Comment by psyke83** [comment]
The major bugs we discussed are all gone with 5002 and 5003 together, but please don't feel pressured to compromise the stability of a major release.

**Comment by twinaphex** [comment]
I trust that you have something ready which can solve most of the issues which were present before.

So go ahead and send us a full PR so that we can merge it in time for 1.6.0.

**Comment by psyke83** [comment]
Since the initial fix was reverted, I've merged all fixes into one PR: https://github.com/libretro/RetroArch/pull/5006

The major bugs are solved in this as far as I can see, so it's ready for merging before or after the release, at your discretion. Thanks!

**Comment by retro-wertz** [comment]
i have tested the new PR and i can say that pipeline OFF seems to be working now, but this probably needs tweaking, for the menu color seems to be more apparent(forgive the lack of color description terminology here) like if choosing blue its bluer than before. see below:

XMB Trans:
![xmbtrans](https://cloud.githubusercontent.com/assets/23315382/26665758/63fded76-46ce-11e7-96c0-bfb285e6311c.jpg)
![xmbmenu](https://cloud.githubusercontent.com/assets/23315382/26665775/835c53c4-46ce-11e7-94fb-464e4a32b90e.jpg)

XMB Retroarch default (latest 1.6.0)
![default](https://cloud.githubusercontent.com/assets/23315382/26665768/77d4653c-46ce-11e7-8b14-da34414cd4b8.jpg)
![menudefault](https://cloud.githubusercontent.com/assets/23315382/26665779/8b11156e-46ce-11e7-8633-454ea90c1e18.jpg)

With regards to the graphics corruption, this seems to be an RA issue since it still happens on latest core with the same test method, its just not noticeble due to menu not transparent when closing core.

**Comment by psyke83** [comment]
I'm just a bit confused about your screenshots... it shouldn't be possible to see transparent menus with the shader pipeline off and non-plain colour theme on the 1.6.0 build. Regardless, the corruption issue was introduced sometime after the 1.5.0 build - but isn't related to any of my changes. I find the corruption rather annoying too, so I'm trying to hunt it down.

With my patch applied I do notice a difference in opacity between shader pipeline off and on (more opaque when off). I'm trying to see if there's a way to bring them to parity, but in the meantime, you can simply change the "Menu Alpha Factor" setting to your preference.

**Comment by retro-wertz** [comment]
latest PR test https://github.com/psyke83/RetroArch/commit/f413b23f0642ca9082245fa0ee89192fa0d9b169 
seems OK now aside from alpha levels. RA doesn't seem to crash anymore when loading and closing core without content but corruption is still there though. 
Tested on Win7x64

**Comment by psyke83** [comment]
retro-wertz,

I closed the pull request based on your feedback, but I'm a bit confused about the cause of the corruption issue (corruption is present after stopping content, reloading a core and then toggling the menu before starting content).

* The official Windows x64 1.6.0 build has no corruption
* My x64 build of the latest master branch has corruption
* My x64 build with my transparency patch has corruption

Can you test another build without my patch to check if it's the same for you?

**Comment by retro-wertz** [comment]
Here is latest from buildbot, fresh/clean retroarch.cfg - Win7x64
![a](https://cloud.githubusercontent.com/assets/23315382/26733869/c3c5034a-47ee-11e7-9b06-8c6e35024ccf.jpg)

**Comment by psyke83** [comment]
Thanks, I've reopened the PR considering that the corruption isn't caused by my patch. I tried building against the v1.6.0 tag and saw the same issue, so I think it's related to the toolchain or default build configuration. Do the release builds use a different configuration to nightlies?

**Comment by Kivutar** [comment]
I've merged the PR. If everybody agrees, we can close this.
